pkg/lib/web/generated_asset.go

```diff
@@ -0,0 +1,209 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync/atomic"
+
+	"gopkg.in/fsnotify.v1"
+
+	"github.com/authgear/authgear-server/pkg/util/log"
+	"github.com/authgear/authgear-server/pkg/util/resource"
+)
+
+type LocalFSLogger struct{ *log.Logger }
+
+type GeneratedAssetDescriptor struct {
+	Logger LocalFSLogger
+
+	manifest atomic.Value
+	watcher  *fsnotify.Watcher
+}
+
+type ManifestContext struct {
+	content map[string]string
+}
+
+var _ resource.Descriptor = &GeneratedAssetDescriptor{}
+
+func NewGeneratedAssetDescriptor() *GeneratedAssetDescriptor {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	var manifest atomic.Value
+	manifestContent, err := LoadManifest()
+	if err != nil {
+		fmt.Println(err)
+	}
+	manifest.Store(&ManifestContext{
+		content: manifestContent,
+	})
+
+	s := &GeneratedAssetDescriptor{
+		manifest: manifest,
+		watcher:  watcher,
+	}
+
+	go s.watch()
+
+	manifestPath := "resources/authgear/static/manifest.json"
+	if err = watcher.Add(manifestPath); err != nil {
+		fmt.Println(err)
+	}
+
+	return s
+}
+
+func LoadManifest() (map[string]string, error) {
+	jsonFile, err := os.Open("resources/authgear/static/manifest.json")
+	if err != nil {
+		return nil, err
+	}
+	defer jsonFile.Close()
+
+	byteValue, _ := ioutil.ReadAll(jsonFile)
+
+	var result map[string]string
+	json.Unmarshal([]byte(byteValue), &result)
+
+	return result, nil
+}
+
+func (d *GeneratedAssetDescriptor) watch() {
+	for {
+		select {
+		case event, ok := <-d.watcher.Events:
+			if !ok {
+				return
+			}
+			if event.Op&fsnotify.Write != fsnotify.Write {
+				break
+			}
+			d.Logger.
+				WithField("file", event.Name).
+				Info("change detected, reloading...")
+
+			if err := d.reload(); err != nil {
+				d.Logger.
+					WithError(err).
+					WithField("file", event.Name).
+					Error("reload failed")
+			}
+
+		case err, ok := <-d.watcher.Errors:
+			if !ok {
+				return
+			}
+			d.Logger.WithError(err).Fatal("Watcher failed")
+		}
+	}
+}
+
+func (d *GeneratedAssetDescriptor) reload() error {
+	newManifest, err := LoadManifest()
+	if err != nil {
+		return err
+	}
+
+	manifestCtx := &ManifestContext{
+		content: newManifest,
+	}
+	d.manifest.Store(manifestCtx)
+	return nil
+}
+
+func (d *GeneratedAssetDescriptor) MatchResource(resourcePath string) (*resource.Match, bool) {
+	manifest := d.manifest.Load().(*ManifestContext).content
+
+	key := strings.TrimPrefix(resourcePath, StaticAssetResourcePrefix)
+	if IsSourceMapPath(key) {
+		key = strings.TrimSuffix(key, ".map")
+	}
+
+	if _, ok := manifest[key]; ok {
+		return &resource.Match{}, true
+	}
+	return nil, false
+}
+
+func (d *GeneratedAssetDescriptor) FindResources(fs resource.Fs) ([]resource.Location, error) {
+	manifest := d.manifest.Load().(*ManifestContext).content
+	locations := make([]resource.Location, 0)
+
+	for _, value := range manifest {
+		location := resource.Location{
+			Fs:   fs,
+			Path: fmt.Sprintf("static/%s", value),
```

Comment:
Use [path.Join](https://pkg.go.dev/path#Join) to build path. We do not use [filepath.Join](https://pkg.go.dev/path/filepath#Join) because we always deal with forward-slash path.